types/index.d.ts

```diff
@@ -23,6 +23,17 @@ import { DefaultParams, Params, Match, MatcherFn } from "./matcher";
 export * from "./matcher";
 export * from "./use-location";
 
+type StripAsterisk<Str extends string> =
+  Str extends `${infer Start}*` ? Start : Str;
+
+type ExtractRouteParams<T extends string> = string extends T
+  ? Record<string, string>
+  : T extends `${infer Start}:${infer Param}/${infer Rest}`
+  ? { [k in StripAsterisk<Param> | keyof ExtractRouteParams<Rest>]: string }
+  : T extends `${infer Start}:${infer Param}`
+  ? { [k in StripAsterisk<Param>]: string }
+  : {};
+
 /*
```

Comment:
great job! 👍

`wouter` has a great expansion mechanism, such as [`custom your own matcher`](https://github.com/molefrog/wouter#using-a-path-to-regexp-based-matcher).

so, the `ExtractRouteParams ` should cover as many cases as possible. I am not particularly good at using typescript, i found the [`react-router@5.1.16`](https://unpkg.com/@types/react-router@5.1.16/index.d.ts) has a complete definition, are we missing something?

```js
export type ExtractRouteOptionalParam<T extends string, U = string | number | boolean> = T extends `${infer Param}?`
    ? { [k in Param]?: U }
    : T extends `${infer Param}*`
    ? { [k in Param]?: U }
    : T extends `${infer Param}+`
    ? { [k in Param]: U }
    : { [k in T]: U };

export type ExtractRouteParams<T extends string, U = string | number | boolean> = string extends T
    ? { [k in string]?: U }
    : T extends `${infer _Start}:${infer ParamWithOptionalRegExp}/${infer Rest}`
    ? ParamWithOptionalRegExp extends `${infer Param}(${infer _RegExp})`
      ? ExtractRouteOptionalParam<Param, U> & ExtractRouteParams<Rest, U>
      : ExtractRouteOptionalParam<ParamWithOptionalRegExp, U> & ExtractRouteParams<Rest, U>
    : T extends `${infer _Start}:${infer ParamWithOptionalRegExp}`
    ? ParamWithOptionalRegExp extends `${infer Param}(${infer _RegExp})`
      ? ExtractRouteOptionalParam<Param, U>
      : ExtractRouteOptionalParam<ParamWithOptionalRegExp, U>
    : {};
```

Reply:
This looks good, I'll slot this in when I have the time

Reply:
👍

Reply:
It's in!

Reply:
@molefrog 👀